### PR TITLE
Minor fixes

### PR DESCRIPTION
**Pull request by NiLuJe**
* More robust "plugged in" detection
* Forma/H2O thumbnail detection tweaks

### CODE REVIEW

File: kobo-uncaged/kutypes.go [hunk 1]
```diff
@@ -143,18 +144,22 @@ func (d koboDevice) CoverSize() (fullCover, libFull, libGrid image.Point) {
...
+		fc = image.Pt(1080, 1429)
+		lf = image.Pt(355, 473)
+		lg = image.Pt(149, 198)
 	case auraOne, auraOneLE:
 		fc = image.Pt(1404, 1872)
 		lf = image.Pt(355, 473)
 		lg = image.Pt(149, 198)
-	case forma:
+	case forma, forma32gb:
 		fc = image.Pt(1440, 1920)
```
Comment (kobothingy) [inline kobo-uncaged/kutypes.go]: Full HD is better.

```suggestion
		fc = image.Pt(1920, 1080)
```

File: scripts/run-ku.sh [hunk 2]
```diff
@@ -13,7 +13,7 @@ KU_TMP_DIR="$2"
 
 # Abort if the device is currently plugged in, as that's liable to confuse Nickel into actually starting a real USBMS session!
 # Which'd probably ultimately cause a crash with our shenanigans...
-if [ "$(cat /sys/devices/platform/pmic_battery.1/power_supply/mc13892_bat/status)" = "Charging" ]; then
+if [ $(cat /sys/devices/platform/pmic_battery.1/power_supply/mc13892_charger/online) -ne 0 ]; then
     # Sleep a bit to lose the race with Nickel's opening of our image
```
Comment (kobothingy) [inline scripts/run-ku.sh]: Why not win it?

File: kobo-uncaged/kutypes.go [hunk 3]
```diff
@@ -143,18 +144,22 @@ func (d koboDevice) CoverSize() (fullCover, libFull, libGrid image.Point) {
...
-		lf = image.Pt(355, 473)
-		lg = image.Pt(149, 198)
+		lf = image.Pt(398, 530)
+		lg = image.Pt(167, 223)
```
Comment (shermp) [inline kobo-uncaged/kutypes.go]: Are you sure about this?

If so, you might want to let davidfor know to (potentially) update the Calibre driver, as that is where I got the values.

Comment (geek1011) [reply kobo-uncaged/kutypes.go]: The sizes are at 0x63c848 in fw 4.14.12777. I'll check them to make sure.

Comment (geek1011) [reply kobo-uncaged/kutypes.go]: Here's the CFG of you want to have a look: [imagesize.pdf](https://s3.geek1011.net/temp/imagesize.pdf?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=admin%2F20190520%2F%2Fs3%2Faws4_request&X-Amz-Date=20190520T221456Z&X-Amz-Expires=604800&X-Amz-SignedHeaders=host&X-Amz-Signature=176d4baf2084a36b6ebc24932db0ccff5f0e246eaa6252391a1a538954226a2f)

Comment (geek1011) [reply kobo-uncaged/kutypes.go]: And with the sizes in decimal: [imagesizenum.pdf](https://s3.geek1011.net/temp/imagesizenum.pdf?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=admin%2F20190520%2F%2Fs3%2Faws4_request&X-Amz-Date=20190520T221735Z&X-Amz-Expires=604800&X-Amz-SignedHeaders=host&X-Amz-Signature=df30daac962ed02462ea4cec5d76528b0eba4ce12aeda7b365298b561502cc9b)

**Comment by shermp** [comment]
Interesting. I wonder why the Calibre driver sets a different resolution per device?

Calling @davidfor do you have any insights?

**Comment by geek1011** [comment]
Whoever came up with those probably tried putting a book on the device and looked at the resulting values, but they were inaccurate due to the aspect ratio preservation.

**Comment by geek1011** [comment]
Based on this info, I'd probably suggest simplifying the cover size stuff and changing the cover resize method to bilinear using the rez library (it is the second-fastest in my benchmarks, and it matches the Kobo output).

File: kobo-uncaged/kutypes.go (hunk 3 above)
Comment (NiLuJe) [reply kobo-uncaged/kutypes.go]: Yes, the Calibre values are inaccurate ;).

I have a [pending](https://github.com/NiLuJe/calibre/commit/ed5478cbb547fbfa81272f084554e16068fb35bf) patch in the works.

I was just not happy with the state of dithering in both Calibre & Nickel, but things have changed a bit recently, and I can probably just port the IM ordered dither code again ;p.

Comment (NiLuJe) [reply kobo-uncaged/kutypes.go]: For the record, I came up with those both by feeding Nickel an ePub with a cover at the perfect AR, and by checking what it did for store-bought KePubs with more-or-less correct AR.

Comment (geek1011) [reply kobo-uncaged/kutypes.go]: I wonder how the values I found come into play though.

Comment (NiLuJe) [reply kobo-uncaged/kutypes.go]: I *may* also have been computing the width manually so it *actually* honors AR (and double-checking that this actually makes sense once rendered in the final widget).

Comment (geek1011) [reply kobo-uncaged/kutypes.go]: I see how that *may* have changed the x values to be *slightly* off from the actual, but can still count as *mostly* correct :smiley:.

Comment (NiLuJe) [reply kobo-uncaged/kutypes.go]: I obviously could only do that for devices I actually have, so, H2O & Forma ;).

Comment (NiLuJe) [reply kobo-uncaged/kutypes.go]: And things have probably changed since then, since a random KePub I just checked on the Forma is 355x555 for the N3_LIBRARY_FULL (it appears to be *trying* to honor the original cover's AR @ 921x1440 against a height of 555, which more or less matches what I was doing with my manual computations, except apparently with lower height bounds at the time ^^).

Comment (NiLuJe) [reply kobo-uncaged/kutypes.go]: I also checked that the full-screen size was often just plain wrong, Nickel's or KePubs, given the actual viewport and what Nickel does when displaying the image (that was actually my main focus at the time: getting bit-exact results from a properly encoded& dithering PNG full-screen cover).

Comment (geek1011) [reply kobo-uncaged/kutypes.go]: If what you said is correct, I misinterpreted Qt::KeepAspectRatioByExpanding for Qt::KeepAspectRatio, as what you said means it shrinks with the size being the minimum, not the maximum.

Comment (NiLuJe) [reply kobo-uncaged/kutypes.go]: Random one-liner to check what's on your device, for posterity's sake:

`find /mnt/onboard/.kobo-images/ -name '*N3_LIBRARY_FULL.parsed' | grep -v file___ | sed -e 's/ /\\ /g' | xargs identify`

Comment (geek1011) [reply kobo-uncaged/kutypes.go]: So, I wonder whether we should go with your calculations or try to reproduce Kobo's behaviour exactly with the values below, using bilinear scaling, and reimplementing Qt's Qt::KeepAspectRatioByExpanding mode.

Comment (NiLuJe) [reply kobo-uncaged/kutypes.go]: Extremely unscientific results of that:

I have a *lot* of stuff scaled around a height of x530, and quite a bit of stuff scaled around a width of 355x

Comment (NiLuJe) [reply kobo-uncaged/kutypes.go]: Which, hey, match the values you pulled from Nickel :D, yay?

(A lot of this is from fairly old files, though. The single KePub that ended up @ x555 is brand new, on the other hand).

Comment (geek1011) [reply kobo-uncaged/kutypes.go]: Mine are all 530 (but I normalize every book I own, so...).

Comment (geek1011) [reply kobo-uncaged/kutypes.go]: Yep, they match. *Why* do I keep posting at the exact same time as you! It's almost the 8th time today.

Comment (NiLuJe) [reply kobo-uncaged/kutypes.go]: @geek1011: Given the general trend, you may have been right on the Qt scaling constraints, since 355x is far more common than that single x555 ;).

Comment (NiLuJe) [reply kobo-uncaged/kutypes.go]: Out of my 614 thumbnails, 541 are scaled against x530, and 48 against 355x

Comment (NiLuJe) [reply kobo-uncaged/kutypes.go]: On a personal note, I tend to prefer Qt::KeepAspectRatioByExpanding's behavior, and IIRC, the actual widget has a crapload of leeway, so this should never be an issue.

Comment (geek1011) [reply kobo-uncaged/kutypes.go]: I think I prefer it too. I propose to switch to the values from the firmware below, and reproduce `Qt::KeepAspectRatioByExpanding` in the resize code.

Comment (geek1011) [reply kobo-uncaged/kutypes.go]: @shermp?

**Comment by NiLuJe** [comment]
TL;DR: Yeah, that probably makes sense!

Bonus point: since UNCaGED doesn't actually generate the full-screen cover for performance reasons, this would allow one to chuck the whole device detection out the window ;).

File: kobo-uncaged/kutypes.go (hunk 3 above)
Comment (shermp) [reply kobo-uncaged/kutypes.go]: Fine by me.

I'm not as picky as @NiLuJe is when it comes to cover/thumbnails quality. So doesn't really bother me what approach is taken.

Comment (geek1011) [reply kobo-uncaged/kutypes.go]: OK. If NiLuJe wants to implement this, he can, or else I'll do it later tonight or tomorrow.

Comment (NiLuJe) [reply kobo-uncaged/kutypes.go]: I'll leave the actual Go heavy lifting to you guys, as I'm mostly a tourist in that area ;).

**Comment by shermp** [comment]
> TL;DR: Yeah, that probably makes sense!
> 
> Bonus point: since UNCaGED doesn't actually generate the full-screen cover for performance reasons, this would potentially allow one to chuck the whole device detection out the window ;).
> 
> (... provided we figure out why Nickel's refusing to generate said full-screen version in my latest tests, c.f., [#3 (comment)](https://github.com/shermp/Kobo-UNCaGED/issues/3#issuecomment-494073731))

Still need a model name, but no reason why that couldn't be passed in on the command line from the cli version of FBInk

**Comment by NiLuJe** [comment]
`eval $(./fbink -qe)` will net you a `$deviceName` variable with that ;).

File: kobo-uncaged/kutypes.go (hunk 3 above)
Comment (NiLuJe) [reply kobo-uncaged/kutypes.go]: FWIW, because maths, here's Qt's logic: https://github.com/qt/qtbase/blob/d3b887707d8c2b342153f962814bb0ce2d0db2b7/src/corelib/tools/qsize.cpp

Comment (geek1011) [reply kobo-uncaged/kutypes.go]: Go's not that hard @NiLuJe! You really should learn it.

Comment (NiLuJe) [reply kobo-uncaged/kutypes.go]: I'm good with C & Python for now, but it is fairly non-alien looking, so there's a vague chance I could reasonably get into it one day ;).

(I have, err, fairly harsh initial gut-feeling reactions to a language's syntax :D. I'm still queasy about Lua in that respect, despite years of messing about with WoW UI & KOReader ;)).

Comment (shermp) [reply kobo-uncaged/kutypes.go]: When I first started learning Go, I found many aspects of it more like C than I've found in most other languages I've used. Made it pretty comfortable to start producing something in a couple of days.

Comment (geek1011) [reply kobo-uncaged/kutypes.go]: > (I have, err, fairly harsh initial gut-feeling reactions to a language's syntax :D. I'm still queasy about Lua in that respect, despite years of messing about with WoW UI & KOReader ;))
I still can't get used to Lua. I've attempted to try multiple times, but given up each time. I like C so far (I've only done minor things in it, nothing big except for some embedded stuff), but I find the error handling and string stuff a bit convoluted and tedious (and the lack of portability in libraries drives me nuts too).

I'm mainly a Go, Python, C# (not as much now that I've been using Linux for around 6 years now).

**Comment by geek1011** [comment]
So, what have we decided on doing? Are we going to close this and redo the thumbnail code to match Kobo?

**Comment by NiLuJe** [comment]
I wouldn't mind this going in just for posterity's sake ;D.

The first commit is at the very least completely orthogonal to the thumbnail thingy, and good to go ;).

**Comment by geek1011** [comment]
OK, I agree. The second commit doesn't even necessarily *break* things or decrease quality per say, although it is technically incorrect.

For the resizing, we can either stick with the imaging NearestNeighbor resizing (the fastest), switch to the rez Bilinear resizing (the second fastest by a small amount, but has some issues with gamma loss), or switch to the imaging Bilinear resizing (half the speed, but still 10x faster than the current). What do you think @shermp, @NiLuJe?

**Comment by shermp** [comment]
Yeah, may as well merge this as is. We can sort out the thumbnail stuff in a separate PR.

**Comment by shermp** [comment]
I'll also open a new issue to discuss the cover image/thumbnail stuff. No need to continue cluttering up this PR
